subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/topology_aware_scheduler.go

```diff
@@ -0,0 +1,284 @@
+	lowestLCALevel           CellLevel // level of the lowest LCA
+	// pack pods cross different priorities, or inside each priority. the former is for opportunistic
+	// pod scheduling (stay away from regular pods), because regular pods can avoid preempting opportunistic pods
+	// only among buddy cells. the latter is for intra-VC scheduling, because high-priority can avoid preemption
+	// in the whole cluster view, and hence we can pack pods with different priorities.
+	crossPriorityPack bool
+}
+
+// newTopologyAwareScheduler initializes the scheduler by extracting node-level cells
+// (lower-level if no node-level) from a chain cell list.
+func newTopologyAwareScheduler(ccl ChainCellList, crossPriorityPack bool) *topologyAwareScheduler {
+	var l CellLevel
+	for l = CellLevel(1); l <= CellLevel(len(ccl)); l++ {
+		if ccl[l][0].AtOrHigherThanNode() {
+			break
+		}
+	}
+	clusterView := make(CellList, len(ccl[l]))
+	copy(clusterView, ccl[l])
+	return &topologyAwareScheduler{
+		clusterView:       clusterView,
+		crossPriorityPack: crossPriorityPack}
+}
+
+func (t *topologyAwareScheduler) schedule(
+	podGpuNumbers map[int32]int32,
+	p CellPriority) map[int32][]CellList {
+
+	t.podGpuNumbers = []int32{}
+	for gpuNum, podNum := range podGpuNumbers {
+		for i := int32(0); i < podNum; i++ {
+			t.podGpuNumbers = append(t.podGpuNumbers, gpuNum)
+		}
+	}
+	sortGpuNumbers(t.podGpuNumbers)
+
+	// disable preemption first (reduce preemption)
+	priority := opportunisticPriority
+	t.updateClusterView(priority)
+	// try to fit the pods to a set of nodes
+	success := t.findNodesForPods(priority)
+	// enable preemption if scheduling failed
+	if !success && p > opportunisticPriority {
+		priority = p
+		t.updateClusterView(priority)
+		success = t.findNodesForPods(priority)
+	}
+	if !success {
+		return nil
+	}
+	// find GPUs inside the selected node for each pod
+	selectedNodes := make(CellList, len(t.podGpuNumbers))
+	for i := 0; i < len(t.currentSolutionIndices); i++ {
+		selectedNodes[i] = t.clusterView[t.currentSolutionIndices[i]]
+	}
+	nodeGpus := map[Cell]CellList{}
+	podPlacements := map[int32][]CellList{}
+	for podIndex := 0; podIndex < len(t.podGpuNumbers); podIndex++ {
+		gpuNumber := t.podGpuNumbers[podIndex]
+		node := selectedNodes[podIndex]
+		nodeGpus[node] = t.findGpusInNode(node, gpuNumber, priority, nodeGpus[node])
+		if podPlacements[gpuNumber] == nil {
+			podPlacements[gpuNumber] = []CellList{}
+		}
+		podPlacements[gpuNumber] = append(podPlacements[gpuNumber], t.lowestLCASolution)
+	}
+	return podPlacements
+}
+
+// findNodesForPods finds a set of nodes that can accommodate the GPU requirements of the pods.
+func (t *topologyAwareScheduler) findNodesForPods(p CellPriority) bool {
+	sort.Sort(t.clusterView)
+	t.currentSolutionIndices = make([]int32, len(t.podGpuNumbers))
+	podIndex := 0
+	pickedGpuNum := int32(0)
+	var node Cell
+	for nodeIndex := 0; nodeIndex < len(t.clusterView); {
+		node = t.clusterView[nodeIndex]
```

Comment:
Can you ensure the algo always can find the solution, regardless the iteration order?

For example:
1. clusterView = 2GPU Node, 1GPU Node
2. podGpuNumbers = 1GPU Pod, 2GPU Pod

First 1GPU Pod may allocate to 2GPU Node, but the latter pod cannot be fitted anymore.

Reply:
OT jobs may cannot ensure this.
Will consider in future.

---
In reply to: [332868715](https://github.com/microsoft/pai/pull/3684#discussion_r332868715) [](ancestors = 332868715)